lib/trie/hash.go

```diff
@@ -59,3 +93,134 @@ func (h *Hasher) Hash(n node) (res []byte, err error) {
 
 	return res, err
 }
+
+// encode is the high-level function wrapping the encoding for different node types
+// encoding has the following format:
+// NodeHeader | Extra partial key length | Partial Key | Value
+func (h *Hasher) encode(n node) ([]byte, error) {
+	switch n := n.(type) {
+	case *branch:
+		return h.encodeBranch(n)
+	case *leaf:
+		return h.encodeLeaf(n)
+	case nil:
+		return []byte{0}, nil
+	}
+
+	return nil, nil
+}
+
+// encodeBranch encodes a branch with the encoding specified at the top of this package
+func (h *Hasher) encodeBranch(b *branch) ([]byte, error) {
+	if !b.dirty && b.encoding != nil {
+		return b.encoding, nil
+	}
+	h.tmp.Reset()
+
+	encoding, err := b.header()
+	h.tmp = append(h.tmp, encoding...)
+	if err != nil {
+		return nil, err
+	}
+
+	h.tmp = append(h.tmp, nibblesToKeyLE(b.key)...)
+	h.tmp = append(h.tmp, common.Uint16ToBytes(b.childrenBitmap())...)
+
+	if b.value != nil {
+		buffer := bytes.Buffer{}
+		se := scale.Encoder{Writer: &buffer}
+		_, err = se.Encode(b.value)
+		if err != nil {
+			return h.tmp, err
+		}
+		h.tmp = append(h.tmp, buffer.Bytes()...)
+	}
+
+	if h.parallel {
+		wg, _ := errgroup.WithContext(context.Background())
+		resBuff := make([][]byte, 16)
+		for i := 0; i < 16; i++ {
+			func(i int) {
+				wg.Go(func() error {
+					child := b.children[i]
+					if child == nil {
+						return nil
+					}
+
+					hasher := NewHasher(false)
+					defer returnHasherToPool(hasher)
+
+					encChild, err := hasher.Hash(child)
+					if err != nil {
+						return err
+					}
+
+					scEncChild, err := scale.Encode(encChild)
+					if err != nil {
+						return err
+					}
```

Comment:
this code is duplicated below, create a separate func for it?

Reply:
Done